crates/datastore/src/system_tables.rs

```diff
@@ -108,6 +110,73 @@ pub(crate) const ST_VIEW_ARG_NAME: &str = "st_view_arg";
 /// test suite when adding sequences to system tables.
 pub const ST_RESERVED_SEQUENCE_RANGE: u32 = 4096;
 
+/// Is `table_id` reserved as a system table?
+pub fn table_id_is_reserved(table_id: TableId) -> bool {
+    table_id.0 <= ST_RESERVED_SEQUENCE_RANGE
+}
+
+/// For a `row` in the table `table_id`, is `row` a schema meta-descriptor for a reserved system table?
+///
+/// Returns true e.g. for the `st_table` row `{ table_id: ST_VIEW_ID, table_name: "st_view", .. }`,
+/// but false e.g. for the `st_table` row `{ table_id: ST_RESERVED_SEQUENCE_RANGE + 1, table_name: "some_user_table", .. }`.
+/// Also returns false if `table_id` is not a system table.
+pub fn is_built_in_meta_row(table_id: TableId, row: &ProductValue) -> Result<bool, anyhow::Error> {
+    fn to_typed_row<T: DeserializeOwned>(row: &ProductValue) -> Result<T, anyhow::Error> {
+        T::deserialize(ValueDeserializer::new(AlgebraicValue::Product(row.clone())))
+            .map_err(|e| anyhow::anyhow!("Failed to deserialize {row:?} as {}: {e:?}", std::any::type_name::<T>()))
+    }
+
+    Ok(match table_id {
+        ST_TABLE_ID => {
+            let row: StTableRow = to_typed_row(row)?;
+            table_id_is_reserved(row.table_id)
+        }
+        ST_COLUMN_ID => {
+            let row: StColumnRow = to_typed_row(row)?;
+            table_id_is_reserved(row.table_id)
+        }
+        ST_SEQUENCE_ID => {
+            let row: StSequenceRow = to_typed_row(row)?;
+            table_id_is_reserved(row.table_id)
+        }
+        ST_INDEX_ID => {
+            let row: StIndexRow = to_typed_row(row)?;
+            table_id_is_reserved(row.table_id)
+        }
+        ST_CONSTRAINT_ID => {
+            let row: StConstraintRow = to_typed_row(row)?;
+            table_id_is_reserved(row.table_id)
+        }
+        ST_MODULE_ID => false,
+        ST_CLIENT_ID => false,
+        ST_VAR_ID => false,
+        ST_SCHEDULED_ID => {
+            // We don't have any scheduled system tables as of writing (pgoldman 2025-12-16),
+            // but no harm in future-proofing.
+            let row: StScheduledRow = to_typed_row(row)?;
+            table_id_is_reserved(row.table_id)
+        }
+        ST_ROW_LEVEL_SECURITY_ID => {
+            // We don't install any RLS rules on system tables automatically, but users can.
+            // This means that `st_row_level_security` rules are never system meta-descriptors,
+            // in the sense that if they exist, they come from users.
+            false
+        }
+        ST_CONNECTION_CREDENTIALS_ID => false,
+        // We don't define any system views, so none of the view-related tables can be system meta-descriptors.
+        ST_VIEW_ID => false,
+        ST_VIEW_PARAM_ID => false,
+        ST_VIEW_COLUMN_ID => false,
+        ST_VIEW_SUB_ID => false,
+        ST_VIEW_ARG_ID => false,
+        TableId(17..ST_RESERVED_SEQUENCE_RANGE) => {
```

Comment:
Could we replace the magic number with a computed `const` item, or use a half-open range?

```suggestion
        TableId(..ST_RESERVED_SEQUENCE_RANGE) => {
```